### PR TITLE
 [reviewed-proto-renderer Part 0]: Base HTML Package

### PR DESCRIPTION
**Pull request by Marcial1234**
Part of #247 

Proto templates on next PR

### CODE REVIEW

File: claat/proto-renderer/error_utils.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package genrenderer
+
+import (
+	"errors"
+)
+
+func AssertError(el interface{}) error {
```
Comment (cassierecher) [inline claat/proto-renderer/error_utils.go]: Comment explaining what this is, what the arg should be, and why its necessary

File: claat/proto-renderer/html/render.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	"go/build"
+	"io"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/googlecodelabs/tools/claat/proto-renderer"
+)
+
+const (
```
Comment (cassierecher) [inline claat/proto-renderer/html/render.go]: Don't use () form if there's just one

File: claat/proto-renderer/html/render.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	defer func() {
+		r := recover()
+		if r != nil {
+			out = nil
+			err = genrenderer.AssertError(r)
+		}
+	}()
+
+	out = strings.NewReader(genrenderer.ExecuteTemplate(el, t))
+	// Compile time error if we omit this line even with named returns
```
Comment (cassierecher) [inline claat/proto-renderer/html/render.go]: Can drop this comment, I was mistaken

File: claat/proto-renderer/html/render_test.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	"testing"
+)
+
+type encapsulatedTest struct {
+	in  interface{}
+	out string
+	ok  bool
+}
+
+type unsupportedProto struct{}
```
Comment (cassierecher) [inline claat/proto-renderer/html/render_test.go]: Comment explaining this

File: claat/proto-renderer/html/render_test.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+		{unsupportedProto{}, "", false},
+	}
+
+	for _, tc := range tests {
+		runEncapsulatedTest(tc, t)
+	}
+}
+
+func runEncapsulatedTest(test encapsulatedTest, t *testing.T) {
+	// Check wheather template failed to render by checking for panic
```
Comment (cassierecher) [inline claat/proto-renderer/html/render_test.go]: nit: whether

File: claat/proto-renderer/html/render_test.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+		if err != nil && test.ok {
+			t.Errorf("\nRender(\n\t%#v\n) = %#v\nPanic occured:\n\t%#v\n(false negative)", test, test.out, err)
+		}
+
+		if err == nil && !test.ok {
+			t.Errorf("\nRender(\n\t%#v\n) = %#v\nWant panic\n(false positive)", test, test.out)
+		}
+	}(test)
+
+	o, _ := Render(test.in)
```
Comment (cassierecher) [inline claat/proto-renderer/html/render_test.go]: Don't discard the error, check for it. "this should never happen" tends to happen quite a lot IME

File: claat/proto-renderer/html/render.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+const (
+	tmplsRltvDir = "src/github.com/googlecodelabs/tools/claat/proto-renderer/html/templates/*"
+)
+
+var (
+	tmplsAbsDir = filepath.Join(build.Default.GOPATH, tmplsRltvDir)
+	t           = template.Must(template.New("html").ParseGlob(tmplsAbsDir))
+)
+
+// Render returns the rendered HTML representation of a devrel_tutorial proto,
```
Comment (cassierecher) [inline claat/proto-renderer/html/render.go]: Remove "devrel_"

Comment (Marcial1234) [reply claat/proto-renderer/html/render.go]: If so, it'll it implies it only renders `devrel_tutorial.Tutorial`. Trying to make explicit that it renders _any_ proto from the `devrel_tutorial` package

File: claat/proto-renderer/generic_renderer_test.go [hunk 8]
```diff
@@ -30,31 +22,34 @@ func TestExecuteTemplate(t *testing.T) {
...
-	if test.out != tmplOut {
-		t.Errorf("Expecting:\n\t'%s'\nBut got:\n\t'%s'", test.out, tmplOut)
+	tmplOut = ExecuteTemplate(tc.in, tmpl)
+	// never gets here if above panicked
+	if tc.out != tmplOut {
+		t.Errorf("Expecting:\n\t'%s'\nBut got:\n\t'%s'", tc.out, tmplOut)
 	}
+	// dummy return, using for shared defer scope
+	return tmplOut
 }
```
Comment (Marcial1234) [inline claat/proto-renderer/generic_renderer_test.go]: Errors here will be formatter as follows:

image

File: claat/proto-renderer/html/render_test.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+		}
+	}
+}
+
+// readerToString Making io.Reader more readable for errors
+func readerToString(i io.Reader) string {
+	var b bytes.Buffer
+	b.ReadFrom(i)
+	return b.String()
+}
```
Comment (Marcial1234) [inline claat/proto-renderer/html/render_test.go]: Errors here will be as follows:

image

File: claat/proto-renderer/tests_utils.go [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+// TODO: delete these after next PR
+// SampleProtoTemplate serves as a dummy supported proto before proto redefinitions
+type SampleProtoTemplate struct {
+	Value interface{}
+}
+
+// NewSampleProtoTemplate pointer constructor for the above
+func NewSampleProtoTemplate(el interface{}) *SampleProtoTemplate {
+	return &SampleProtoTemplate{Value: el}
+}
```
Comment (Marcial1234) [inline claat/proto-renderer/tests_utils.go]: "Small PRs" Overhead

File: claat/proto-renderer/generic_renderer.go [hunk 11]
```diff
@@ -3,15 +3,25 @@ package genrenderer
...
 
+// TODO: update to be proto dependent on next PR
+func templateName(el interface{}) string {
+	switch el.(type) {
+	// This PR only comment: returns 'sampleProtoTemplate' to not break
+	// 'ExecuteTemplate' tests
+	case *SampleProtoTemplate:
+		return "sampleProtoTemplate"
+	}
+	return fmt.Sprintf("TYPE NOT SUPPORTED: %T, Val: %#v", el, el)
```
Comment (cassierecher) [inline claat/proto-renderer/generic_renderer.go]: lowercase is fine

**Pull request merged**